src/index/sparse/sparse_inverted_index.h

```diff
@@ -187,68 +205,169 @@ class InvertedIndex : public BaseInvertedIndex<T> {
             }
             add_row_to_index(raw_data_[i], i);
         }
-
         return Status::success;
     }
 
-    // Non zero drop ratio is only supported for static index, i.e. data should
-    // include all rows that'll be added to the index.
+    // memory in reader must be guaranteed to be valid during the lifetime of this object.
     Status
-    Train(const SparseRow<T>* data, size_t rows, float drop_ratio_build) override {
-        if (drop_ratio_build == 0.0f) {
-            return Status::success;
-        }
-        // TODO: maybe i += 10 to down sample to speed up.
-        size_t amount = 0;
-        for (size_t i = 0; i < rows; ++i) {
-            amount += data[i].size();
-        }
-        if (amount == 0) {
-            return Status::success;
-        }
-        std::vector<T> vals;
-        vals.reserve(amount);
+    PrepareMmap(MemoryIOReader& reader, size_t rows, int map_flags) {
+        const auto initial_reader_location = reader.tellg();
+        const auto nnz = (reader.remaining() - (rows * sizeof(size_t))) / SparseRow<T>::element_size();
+
+        // count raw vector idx occurrences
+        std::unordered_map<table_t, size_t> idx_counts;
         for (size_t i = 0; i < rows; ++i) {
-            for (size_t j = 0; j < data[i].size(); ++j) {
-                vals.push_back(fabs(data[i][j].val));
+            size_t row_nnz;
+            readBinaryPOD(reader, row_nnz);
+            if (row_nnz == 0) {
+                continue;
+            }
+            for (size_t j = 0; j < row_nnz; ++j) {
+                table_t idx;
+                readBinaryPOD(reader, idx);
+                idx_counts[idx]++;
+                // skip value
+                reader.advance(sizeof(T));
             }
         }
-        auto pos = vals.begin() + static_cast<size_t>(drop_ratio_build * vals.size());
-        // pos may be vals.end() if drop_ratio_build is 1.0, in that case we use
-        // the largest value as the threshold.
-        if (pos == vals.end()) {
-            pos--;
+        // reset reader to the beginning
+        reader.seekg(initial_reader_location);
+
+        auto raw_data_size = rows * sizeof(typename decltype(raw_data_)::value_type);
```

Comment:
maybe, it would be nice if variables could be called as `raw_data_size_in_bytes` rather than `raw_data_size`, because the default convension is that `*_size` is measured in elements, not in bytes.

No big deal, but it just improves the readibility of the code.